TypeSystem/flow-analysis/promotion_via_assignment_A03_t03.dart

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion We say that a variable `x` is promotable via assignment of an
+/// expression of type `T` given variable model `VM` if
+/// - `VM = VariableModel(declared, promoted, tested, assigned, unassigned,
+///     captured)`
+/// - and captured is false
+/// - and `S` is the current type of `x` in `VM`
+/// - and `T <: S` and not `S <: T`
+/// - and `T` is a type of interest for `x` in `tested`
+///
+/// @description Checks that if `T <: S` and `S <: T` then promotion via
+/// assignment is not performed. Test top types.
+/// @author sgrekhov22@gmail.com
+/// @issue 60620
+
+import '../../Utils/static_type_helper.dart';
+
+typedef Void = void;
+
+test1() {
+  Object? v = 1 > 2 ? null : Object();
+  if (v is dynamic) {} // ignore: unnecessary_type_check
+  v = (1 > 2 ? null : Object()) as dynamic;
+  // `Object?` <: `dynamic` and `dynamic` <: `Object?`.
+  // Therefore `v` is not promoted to `dynamic`.
+  v.testDynamic;
+//  ^^^^^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+}
+
+test2() {
+  Object? v = 1 > 2 ? null : Object();
+  if (v is Void) {} // ignore: unnecessary_type_check
+  v = 1 > 2 ? null : Object(); // can't assign `void`, let's assign subtype of `void`
+  // `Object?` <: `void` and `void` <: `Object?`.
+  // Therefore `v` is not promoted to `void`.
+  v.expectStaticType<Exactly<Object?>>();
```

Comment:
That's a weird way to verify that the static type is `Object?` rather than `void`. ;-)

The point is that they are mutual subtypes, so the `expectStaticType` test would just say yes ("`v` has type `void` so we can confirm that it has type `Object?`!").

However, if `v` actually has type `void` then it is an error to invoke `expectStaticType` on it. So we _will_ get an error if `v` has type `void`, but for a reason which is not readable.

Perhaps just do `v.hashCode` in order to get the error if the type is `void`?

Reply:
Thx! ;-)